File: crates/lae-core/src/install/hypr.rs
//! Hyprland integration install / uninstall.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use chrono::Utc;
use serde_json::{json, Value};

use crate::config::LaeConfig;
use crate::error::{LaeError, Result};
use crate::install::backup::{self, backup_timestamp};
use crate::install::manifest::{self, Manifest};
use crate::install::path_link;
use crate::install::reload;
use crate::install::wrapper;
use crate::xdg::{config_home, ensure_parent, expand, user_bin_dir};

#[derive(Debug, Clone)]
pub struct InstallHyprOptions {
    pub dry_run: bool,
    pub workspace_root: Option<PathBuf>,
}

impl Default for InstallHyprOptions {
    fn default() -> Self {
        Self {
            dry_run: false,
            workspace_root: None,
        }
    }
}

pub fn install_hypr_status(cfg: &LaeConfig) -> Result<Value> {
    let m = manifest::load_manifest(&cfg.install_hypr_share_dir, "hypr")?;
    let bindings = cfg.install_hypr_share_dir.join("hypr/bindings.conf");
    let tui_helper = cfg.install_hypr_share_dir.join("bin/lae-task-tui");
    let has_source = cfg.install_hypr_config_path.is_file()
        && fs::read_to_string(&cfg.install_hypr_config_path)
            .map(|s| s.contains("lae-managed"))
            .unwrap_or(false);
    Ok(json!({
        "installed": m.is_some(),
        "bindings_exist": bindings.is_file(),
        "tui_helper_exist": tui_helper.is_file(),
        "source_line_present": has_source,
        "config_path": cfg.install_hypr_config_path,
        "bindings_path": bindings,
    }))
}

pub fn install_hypr(cfg: &LaeConfig, options: &InstallHyprOptions) -> Result<Vec<String>> {
    let share_src = find_share_root(options.workspace_root.as_deref())?.join("hypr");
    let share_dest = cfg.install_hypr_share_dir.join("hypr");
    let backup_dir = cfg
        .install_hypr_share_dir
        .join("install/hypr/backups")
        .join(backup_timestamp());
    let source_line = format!(
        "source = {}  # lae-managed (installed {})",
        cfg.install_hypr_source_line,
        Utc::now().date_naive()
    );

    if options.dry_run {
        return Ok(vec![
            format!("would copy {} → {}", share_src.display(), share_dest.display()),
            format!("would append to {}", cfg.install_hypr_config_path.display()),
            format!("would install {}", cfg.install_hypr_share_dir.join("bin/lae").display()),
            format!(
                "would symlink {} → {}",
                user_bin_dir().join("lae").display(),
                cfg.install_hypr_share_dir.join("bin/lae").display()
            ),
        ]);
    }

    ensure_parent(&share_dest.join("_"))?;
    if share_src.is_dir() {
        for entry in fs::read_dir(&share_src).map_err(|source| LaeError::Read {
            path: share_src.clone(),
            source,
        })? {
            let entry = entry.map_err(|source| LaeError::Read {
                path: share_src.clone(),
                source,
            })?;
            if entry.file_type().map_err(|source| LaeError::Read {
                path: entry.path(),
                source,
            })?.is_file()
            {
                let dest = share_dest.join(entry.file_name());
                ensure_parent(&dest)?;
                fs::copy(entry.path(), &dest).map_err(|source| LaeError::Write {
                    path: dest,
                    source,
                })?;
            }
        }
    }

    cleanup_legacy_menu(cfg)?;
    let rust_bin = build_and_install_cli(cfg, options.workspace_root.as_deref())?;
    path_link::install_path_symlink(cfg, &rust_bin)?;
    wrapper::write_install_helpers(cfg)?;

    let config_path = &cfg.install_hypr_config_path;
    let mut backed_up = Vec::new();
    let mut modified = false;
    if config_path.is_file() {
        backup::backup_file(config_path, &backup_dir)?;
        backed_up.push(json!({"path": config_path, "backup": config_path.file_name()}));
        let content = fs::read_to_string(config_path).map_err(|source| LaeError::Read {
            path: config_path.clone(),
            source,
        })?;
        if !content.contains("lae-managed") {
            fs::OpenOptions::new()
                .append(true)
                .open(config_path)
                .map_err(|source| LaeError::Write {
                    path: config_path.clone(),
                    source,
                })?
                .write_all(format!("\n{source_line}\n").as_bytes())
                .map_err(|source| LaeError::Write {
                    path: config_path.clone(),
                    source,
                })?;
            modified = true;
        }
    } else {
        ensure_parent(config_path)?;
        fs::write(config_path, format!("{source_line}\n")).map_err(|source| LaeError::Write {
            path: config_path.clone(),
            source,
        })?;
        modified = true;
    }

    let m = Manifest {
        version: 1,
        integration: "hypr".into(),
        installed_at: Utc::now().to_rfc3339(),
        backup_dir: backup_dir.to_string_lossy().into_owned(),
        templates_installed: vec![json!({"from": share_src, "to": share_dest})],
        user_files_backed_up: backed_up,
        user_files_modified: if modified {
            vec![json!({"path": config_path, "actions": [{"type": "append", "line": source_line}]})]
        } else {
            vec![]
        },
        module_kind: None,
    };
    manifest::save_manifest(&cfg.install_hypr_share_dir, &m)?;

    reload::apply_after_hypr()
}

pub fn uninstall_hypr(cfg: &LaeConfig, keep_files: bool) -> Result<Vec<String>> {
    let m = manifest::load_manifest(&cfg.install_hypr_share_dir, "hypr")?
        .ok_or_else(|| LaeError::Other("No lae Hyprland installation found".into()))?;

    let backup_root = PathBuf::from(&m.backup_dir);
    for entry in &m.user_files_backed_up {
        if let (Some(path), Some(backup)) = (entry.get("path"), entry.get("backup")) {
            let src = backup_root.join(backup.as_str().unwrap_or(""));
            let dest = expand(path.as_str().unwrap_or(""));
            if src.is_file() {
                backup::restore_file(&src, &dest)?;
            }
        }
    }

    cleanup_legacy_menu(cfg)?;

    let rust_bin = cfg.install_hypr_share_dir.join("bin/lae");
    let _ = path_link::remove_path_symlink(&rust_bin);

    if !keep_files {
        let hypr_dir = cfg.install_hypr_share_dir.join("hypr");
        if hypr_dir.is_dir() {
            let _ = fs::remove_dir_all(hypr_dir);
        }
    }

    manifest::remove_manifest(&cfg.install_hypr_share_dir, "hypr")?;
    reload::apply_after_hypr()
}

fn cleanup_legacy_menu(cfg: &LaeConfig) -> Result<()> {
    let elephant_link = config_home().join("elephant/menus/lae_tasks.lua");
    if elephant_link.is_symlink() {
        let _ = fs::remove_file(elephant_link);
    }
    let elephant_dir = cfg.install_hypr_share_dir.join("elephant");
    if elephant_dir.is_dir() {
        let _ = fs::remove_dir_all(elephant_dir);
    }
    let menu_json = cfg.install_hypr_share_dir.join("bin/lae-task-menu-json");
    if menu_json.is_file() {
        let _ = fs::remove_file(menu_json);
    }
    Ok(())
}

fn build_and_install_cli(cfg: &LaeConfig, workspace_root: Option<&Path>) -> Result<PathBuf> {
    let bin_dir = cfg.install_hypr_share_dir.join("bin");
    let dest = bin_dir.join("lae");
    ensure_parent(&dest)?;

    let workspace = workspace_root
        .map(Path::to_path_buf)
        .or_else(find_workspace_root)
        .ok_or_else(|| {
            LaeError::Other(
                "could not find Cargo workspace — set LAE_WORKSPACE or run from the repo".into(),
            )
        })?;

    let target_dir = workspace.join("target");
    let release_bin = target_dir.join("release/lae");
    eprintln!("building lae CLI (release)...");
    let status = Command::new("cargo")
        .args([
            "build",
            "-p",
            "lae-cli",
            "--release",
            "--target-dir",
        ])
        .arg(&target_dir)
        .current_dir(&workspace)
        .status()
        .map_err(|e| LaeError::Other(format!("failed to run cargo: {e}")))?;
    if !status.success() {
        return Err(LaeError::Other("cargo build -p lae-cli failed".into()));
    }

    let staging = bin_dir.join("lae.new");
    if staging.exists() {
        let _ = fs::remove_file(&staging);
    }
    fs::copy(&release_bin, &staging).map_err(|source| LaeError::Write {
        path: staging.clone(),
        source,
    })?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = fs::metadata(&staging)
            .map_err(|source| LaeError::Read {
                path: staging.clone(),
                source,
            })?
            .permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&staging, perms).map_err(|source| LaeError::Write {
            path: staging.clone(),
            source,
        })?;
    }
    fs::rename(&staging, &dest).map_err(|source| LaeError::Write {
        path: dest.clone(),
        source,
    })?;
    Ok(dest)
}

fn find_share_root(workspace_root: Option<&Path>) -> Result<PathBuf> {
    if let Some(root) = workspace_root {
        return Ok(root.join("share"));
    }
    find_workspace_root()
        .map(|w| w.join("share"))
        .ok_or_else(|| LaeError::Other("could not find share/ templates".into()))
}

fn find_workspace_root() -> Option<PathBuf> {
    if let Ok(env) = std::env::var("LAE_WORKSPACE") {
        let p = PathBuf::from(env);
        if p.join("Cargo.toml").is_file() {
            return Some(p);
        }
    }
    let mut dir = std::env::current_dir().ok()?;
    loop {
        if dir.join("Cargo.toml").is_file() && dir.join("share/hypr").is_dir() {
            return Some(dir);
        }
        if !dir.pop() {
            break;
        }
    }
    None
}

use std::io::Write;

File: src/tsk/install/hypr.py
"""Hyprland integration install / uninstall."""

from __future__ import annotations

import shutil
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from tsk.core.config import TskConfig, load_config
from tsk.install import backup, manifest
from tsk.install.reload import apply_after_hypr
from tsk.install.wrapper import (
    bin_dir,
    write_tsk_wrapper,
    write_tui_launch_helper,
    write_waybar_helper,
)
from tsk.integrations import hyprland
from tsk.util import xdg


@dataclass
class InstallPlan:
    templates: list[tuple[Path, Path]]
    config_path: Path
    source_line: str
    backup_dir: Path
    already_installed: bool
    reload_actions: list[str] | None = None


def _repo_share() -> Path:
    root = Path(__file__).resolve().parents[3]
    share = root / "share"
    if share.is_dir():
        return share
    return xdg.share_dir()


def plan_install(cfg: TskConfig | None = None) -> InstallPlan:
    cfg = cfg or load_config()
    share_src = _repo_share() / "hypr"
    share_dest = cfg.install_hypr_share_dir / "hypr"
    templates: list[tuple[Path, Path]] = []
    for src in share_src.glob("*"):
        if src.is_file():
            templates.append((src, share_dest / src.name))

    ts = backup.backup_timestamp()
    backup_dir = cfg.install_hypr_share_dir / "install" / "hypr" / "backups" / ts
    existing = manifest.load_manifest(cfg.install_hypr_share_dir, "hypr")
    source_line = (
        f"source = {xdg.expand(cfg.install_hypr_source_line)}  "
        f"# tsk-managed (installed {datetime.now(timezone.utc).date().isoformat()})"
    )
    return InstallPlan(
        templates=templates,
        config_path=cfg.install_hypr_config_path,
        source_line=source_line,
        backup_dir=backup_dir,
        already_installed=existing is not None,
    )


def install_hypr(*, dry_run: bool = False, cfg: TskConfig | None = None) -> InstallPlan:
    cfg = cfg or load_config()
    plan = plan_install(cfg)

    if dry_run:
        return plan

    for src, dest in plan.templates:
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)

    write_tui_launch_helper(cfg)
    write_waybar_helper(cfg)
    tsk_wrapper = write_tsk_wrapper(cfg)

    config_path = plan.config_path
    backed_up: list[dict[str, str]] = []
    modified = False

    if config_path.exists():
        backup.backup_file(config_path, plan.backup_dir)
        backed_up.append({"path": str(config_path), "backup": config_path.name})
        content = config_path.read_text()
        if "tsk-managed" not in content:
            with config_path.open("a") as f:
                f.write("\n" + plan.source_line + "\n")
            modified = True
    else:
        config_path.parent.mkdir(parents=True, exist_ok=True)
        config_path.write_text(plan.source_line + "\n")
        modified = True

    templates_installed = [{"from": str(s), "to": str(d)} for s, d in plan.templates]
    templates_installed.append({"generated": str(tsk_wrapper)})
    templates_installed.append({"generated": str(bin_dir(cfg) / "tsk-task-tui")})
    templates_installed.append({"generated": str(bin_dir(cfg) / "tsk-waybar-json")})

    m = manifest.Manifest(
        integration="hypr",
        backup_dir=str(plan.backup_dir),
        templates_installed=templates_installed,
        user_files_backed_up=backed_up,
        user_files_modified=(
            [
                {
                    "path": str(config_path),
                    "actions": [{"type": "append", "line": plan.source_line}],
                }
            ]
            if modified
            else []
        ),
    )
    manifest.save_manifest(cfg.install_hypr_share_dir, m)

    plan.reload_actions = apply_after_hypr()

    return plan


def uninstall_hypr(*, keep_files: bool = False, cfg: TskConfig | None = None) -> list[str]:
    cfg = cfg or load_config()
    m = manifest.load_manifest(cfg.install_hypr_share_dir, "hypr")
    if m is None:
        raise RuntimeError("No tsk Hyprland installation found")

    backup_root = Path(m.backup_dir)
    for entry in m.user_files_backed_up:
        src = backup_root / entry["backup"]
        dest = Path(entry["path"]).expanduser()
        if src.exists():
            backup.restore_file(src, dest)

    if not keep_files:
        hypr_dir = cfg.install_hypr_share_dir / "hypr"
        if hypr_dir.exists():
            shutil.rmtree(hypr_dir)

    manifest_path = cfg.install_hypr_share_dir / "install" / "hypr" / "manifest.json"
    if manifest_path.exists():
        manifest_path.unlink()

    return apply_after_hypr()


def install_status(cfg: TskConfig | None = None) -> dict:
    cfg = cfg or load_config()
    m = manifest.load_manifest(cfg.install_hypr_share_dir, "hypr")
    bindings = cfg.install_hypr_share_dir / "hypr" / "bindings.conf"
    tui_helper = cfg.install_hypr_share_dir / "bin" / "tsk-task-tui"
    config_path = cfg.install_hypr_config_path
    has_source = False
    if config_path.exists():
        has_source = "tsk-managed" in config_path.read_text()
    return {
        "installed": m is not None,
        "manifest": m.to_dict() if m else None,
        "bindings_exist": bindings.exists(),
        "tui_helper_exist": tui_helper.exists(),
        "source_line_present": has_source,
        "config_path": str(config_path),
        "bindings_path": str(bindings),
    }


def doctor_checks(cfg: TskConfig | None = None) -> list[tuple[str, bool, str]]:
    cfg = cfg or load_config()
    checks: list[tuple[str, bool, str]] = []
    status = install_status(cfg)

    checks.append(
        (
            "Hyprland bindings installed",
            status["bindings_exist"],
            str(cfg.install_hypr_share_dir / "hypr" / "bindings.conf"),
        )
    )
    checks.append(
        (
            "Task manager launcher installed",
            status["tui_helper_exist"],
            str(cfg.install_hypr_share_dir / "bin" / "tsk-task-tui"),
        )
    )
    checks.append(
        (
            "hyprland.conf contains tsk source line",
            status["source_line_present"],
            str(cfg.install_hypr_config_path),
        )
    )

    backup_ok = False
    backup_msg = "no manifest"
    if status["manifest"]:
        backup_dir = Path(status["manifest"]["backup_dir"])
        backup_ok = backup_dir.is_dir() and any(backup_dir.iterdir())
        backup_msg = str(backup_dir)
    checks.append(("Install backup exists", backup_ok, backup_msg))

    from tsk.daemon.server import is_daemon_running

    checks.append(
        (
            "Daemon reachable",
            is_daemon_running(),
            str(xdg.tsk_daemon_socket()),
        )
    )
    checks.append(
        (
            "SUPER+1 runs tsk (not Omarchy workspace)",
            _super_one_is_tsk(),
            "hyprctl binds -j",
        )
    )
    return checks


def _super_one_is_tsk() -> bool:
    if not hyprland.available():
        return False
    try:
        binds = hyprland.hyprctl_json("binds") or []
    except hyprland.HyprlandError:
        return False
    tsk_binds = [
        b
        for b in binds
        if b.get("keycode") == 10
        and b.get("modmask") == 64
        and (
            "tsk-workspace-switch" in str(b.get("arg", ""))
            or "tsk workspace go" in str(b.get("arg", ""))
        )
    ]
    omarchy_binds = [
        b
        for b in binds
        if b.get("keycode") in range(10, 20)
        and b.get("modmask") == 64
        and str(b.get("arg", "")).strip().isdigit()
        and "tsk workspace go" not in str(b.get("arg", ""))
    ]
    return bool(tsk_binds) and not omarchy_binds
